### ISSUE TITLE
`while_let_on_iterator` suggested solution doesn't work in default implementations of traits without explicit `Sized` bound

### ISSUE THREAD

**Issue by mks-h** [rust-lang/rust-clippy#16019]
### Summary

The suggested solution for `while_let_on_iterator` fails to compile when applied in a default trait implementation that doesn't have an explicit `Sized` bound.

### Reproducer

I tried this code:

```rust
pub trait CertainTrait: Iterator<Item = Result<u8, ()>> {
	fn parse(&mut self) {
		let mut data = Vec::new();
		while let Some(r) = self.next() {
			match r {
				Err(_) => {},
				Ok(v) => data.push(v),
			}
		}
	}
}
```

I expected to see this happen:
The suggestion should have possibly used `self.into_iter()`, as that works without additional requirements; Or it simply shouldn't have been suggested.

Instead, this happened:
The suggestion (`for r in self.by_ref()`) didn't compile, and the compiler suggested to add `where Self: Sized` to the function definition, which works.

### Version

```text
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=af6bab9b98cca79a75748f4f8963a7cb
```

**Comment by Jarcho** [comment]
This should just suggest `for r in self` since it's already a mutable reference.

**Comment by mks-h** [comment]
That would indeed work for the playground example, but my real code also calls another method with `&mut self` after the loop. In that case it has to be `for r in &mut *self` (as suggested by the compiler) or otherwise you get:
```
`self` moved due to this implicit call to `.into_iter()`
```

**Comment by abhayiscoding** [comment]
```rust
pub trait CertainTrait: Iterator<Item = Result<u8, ()>> {
    fn parse(&mut self) {
        let mut data = Vec::new();
        for r in &mut *self {
            match r {
                Err(_) => {}
                Ok(v) => data.push(v),
            }
        }
    }
}
```

`for r in &mut *self` to avoid requiring `Sized` and keep `self` usable after the loop.

**Comment by robert-mccausland** [comment]
I would like to have a go at this, am I correct in thinking that we should change the refactor from `iter.by_ref()` to `&mut *iter` when the iterator is not sized? 

Regarding the reborrowing being unnecessary here, reading the code it seems that this is a known limitation of this lint, which could be looked at as a follow up.

**Comment by sergiogiro** [comment]
Hi, I was investigating this issue and have a change that passes tests (and has new tests...), and also eliminates the unnecessary reborrowing.

https://github.com/rust-lang/rust-clippy/pull/16100

It's my first contribution to clippy so maybe it's very far off!

If @robert-mccausland has done any work on it, I'm happy for him to take the issue.

**Comment by robert-mccausland** [comment]
> If @robert-mccausland has done any work on it, I'm happy for him to take the issue.

Please go ahead with your changes, I haven’t submitted anything for this issue.

I do wonder about reverting back to borrowing instead of using `by_ref` for all cases, as it seems that this was intentionally changed to `by_ref` in the past for better readability. However probably best to wait for a member of the clippy team to get back with their thoughts on this.

**Comment by sergiogiro** [comment]
> I do wonder about reverting back to borrowing instead of using by_ref for all cases

Indeed, now @Jarcho advised me in the PR to just change the behavior for only non-sized types.

I've now reworked #16100 in this direction.

**Comment by sergiogiro** [comment]
There's been interaction in the PR, I think it makes sense to assign this to me now.

@rustbot claim
